solidity/contracts/token/TokenBridgeCctp.sol

```diff
@@ -117,42 +126,63 @@ abstract contract TokenBridgeCctp is HypERC20Collateral, AbstractCcipReadIsm {
         return domain.circle;
     }
 
+    // @dev Enforces that the CCTP message source domain and nonce matches the Hyperlane message origin and nonce.
     function verify(
         bytes calldata _metadata,
-        bytes calldata /* _message */
+        bytes calldata _hyperlaneMessage
     ) external returns (bool) {
+        // decode return type of CctpService.getCCTPAttestation
         (bytes memory cctpMessage, bytes memory attestation) = abi.decode(
             _metadata,
             (bytes, bytes)
         );
 
+        bytes29 originalMsg = TypedMemView.ref(cctpMessage, 0);
+
+        uint32 sourceDomain = originalMsg._sourceDomain();
+        require(
+            sourceDomain ==
+                hyperlaneDomainToCircleDomain(_hyperlaneMessage.origin()),
+            "Invalid source domain"
+        );
+
+        uint64 sourceNonce = originalMsg._nonce();
+        require(
+            sourceNonce == uint64(bytes8(_hyperlaneMessage.body().metadata())),
+            "Invalid nonce"
+        );
+
         // Receive only if the nonce hasn't been used before
-        if (messageTransmitter.usedNonces(_messageNonce(cctpMessage)) == 0) {
+        bytes32 sourceAndNonceHash = keccak256(
+            abi.encodePacked(sourceDomain, sourceNonce)
+        );
+        if (messageTransmitter.usedNonces(sourceAndNonceHash) == 0) {
             messageTransmitter.receiveMessage(cctpMessage, attestation);
         }
 
         return true;
     }
 
-    function _transferRemote(
-        uint32 _destination,
-        bytes32 _recipient,
-        uint256 _amount,
-        uint256 _value,
-        bytes memory _hookMetadata,
-        address _hook
-    ) internal virtual override returns (bytes32 messageId) {
-        messageId = super._transferRemote(
-            _destination,
-            _recipient,
-            _amount,
-            _value,
-            _hookMetadata,
-            _hook
+    function _transferFromSender(
+        uint256 _amountOrId
+    ) internal virtual override returns (bytes memory metadata) {
+        HypERC20Collateral._transferFromSender(_amountOrId);
+
+        // decode transferRemote(uint32, bytes32, uint256,...) calldata
+        (uint32 destination, bytes32 recipient, uint256 amount) = abi.decode(
+            msg.data[4:],
+            (uint32, bytes32, uint256)
```

Comment:
this seems a little fragile but im testing the call to `depositForBurn`
otherwise I have to change the abstraction